packages/core/src/types.ts

```diff
@@ -0,0 +1,209 @@
+import { AxiosResponse, CancelTokenSource } from 'axios'
+
+export type Errors = Record<string, string>
+export type ErrorBag = Record<string, Errors>
+
+export type FormDataConvertible =
+  | Array<FormDataConvertible>
+  | Blob
+  | FormDataEntryValue
+  | Date
+  | boolean
+  | number
+  | null
+  | undefined
+
+export enum Method {
+  GET = 'get',
+  POST = 'post',
+  PUT = 'put',
+  PATCH = 'patch',
+  DELETE = 'delete',
+}
+
+export type RequestPayload = Record<string, FormDataConvertible> | FormData
+
+export interface PageProps {
+  [key: string]: unknown
+}
+
+export interface Page<SharedProps = PageProps> {
+  component: string
+  props: PageProps &
+    SharedProps & {
+      errors: Errors & ErrorBag
+    }
+  url: string
+  version: string | null
+
+  // Refactor away
+  scrollRegions: Array<{ top: number; left: number }>
+  rememberedState: Record<string, unknown>
+  resolvedErrors: Errors
+}
+
+export type PageResolver = (name: string) => Component
+
+export type PageHandler = ({
+  component,
+  page,
+  preserveState,
+}: {
+  component: Component
+  page: Page
+  preserveState: PreserveStateOption
+}) => Promise<unknown>
+
+export type PreserveStateOption = boolean | string | ((page: Page) => boolean)
+
+export type Progress = ProgressEvent & { percentage: number }
+
+export type LocationVisit = {
+  preserveScroll: boolean
+}
+
+export type Visit = {
+  method: Method
+  data: RequestPayload
+  replace: boolean
+  preserveScroll: PreserveStateOption
+  preserveState: PreserveStateOption
+  only: Array<string>
+  headers: Record<string, string>
+  errorBag: string | null
+  forceFormData: boolean
+  queryStringArrayFormat: 'indices' | 'brackets'
```

Comment:
Maybe add `'comma'` too, since it's supported by the `qs` library?

<img width="622" alt="image" src="https://user-images.githubusercontent.com/3526974/192133769-ac4ae223-ee9e-4a1c-81c2-cbbc2c158f95.png">
https://github.com/ljharb/qs#stringifying

Reply:
Yes, not necessarily opposed to this, but going to leave this for a subsequent update 👍